blog/2024-08-07-howt-to-keep-mqtt-data-rust/index.md

```diff
@@ -0,0 +1,280 @@
+
+        // Retrieve records as a stream
+        let mut record_stream = pin!(record_stream);
+        while let Some(record) = record_stream.next().await {
+            let record = record.unwrap();
+            println!(
+                "MQTT topic: {}, Record: ts={}, data={}",
+                entry.name,
+                record.timestamp_us(),
+                String::from_utf8_lossy(&record.bytes().await.unwrap()),
+            );
+        }
+    }
+}
+```
+
+We have already covered how to create a client and acquire a bucket. Now let's move on to reading the data:
+
+```rust
+    let entries = bucket.entries().await.unwrap();
+
+    for entry in entries {
+        // Query all records from the entry
+        let mut record_stream = bucket.query(&entry.name).send().await.unwrap();
+
+        // Retrieve records as a stream
+        let mut record_stream = pin!(record_stream);
+        while let Some(record) = record_stream.next().await {
+            let record = record.unwrap();
+            println!(
+                "MQTT topic: {}, Record: ts={}, data={}",
+                entry.name,
+                record.timestamp_us(),
+                String::from_utf8_lossy(&record.bytes().await.unwrap()),
+            );
+        }
+    }
+```
+
+As you can see, this is also very simple. We browse all the entries in the `mqtt` bucket, retrieve all the records from each entry, and then print their timestamps and contents. That's it.
+
+## Best Practices
+
+The example is a simple one and may not cover all of the complexities that you may be faced with in a real-world application. Here are some tips to help you build a strong and efficient IoT application using ReductStore and MQTT:
+
+- Create a ReductStore bucket with a **[FIFO quota](docs/guides/buckets.mdx#quota-type)** to prevent disk overwrites.
+- Use token authentication to protect your data. You can generate an access token using either the **[Web Console](https://github.com/reductstore/web-console)** or the **[CLI Client](https://github.com/reductstore/reduct-cli)**.
+- Map **MQTT5** properties to ReductStore labels. This will make it easier to filter data when **[querying](/docs/guides/data-querying.mdx#using-labels-to-filter-data)** or **[replicating](/docs/guides/data-replication.mdx)**.
+- Use **[Reduct CLI](https://github.com/reductstore/reduct-cli)** for data replication or backup purposes.
+
+## Conclusion
+
+The MQTT protocol and ReductStore are easy to use tools that work well together in Rust. They provide a powerful solution for many applications. Regardless of the size of your project, these tools handle data communication and storage effectively.
+
+To help you understand how to use these tools, we've created an example that shows how they work together. You can see the source code of this example **[here](https://github.com/reductstore/reduct-mqtt-example-rs)**. This example shows how easy and useful it is to use MQTT and ReductStore together.
```

Comment:
Needs to be something else for than [here] for SEO (for example [reduct-mqtt-example-rs]), otherwise it triggers an issue on Seobility (if we use "here" a second time; it will say that multiple "here" anchors link to different URLs) and apparently it's not good.